Gems/Atom/RHI/Code/Include/Atom/RHI/MultiDeviceTransientAttachmentPool.h

```diff
@@ -0,0 +1,119 @@
+
+        //! Called when an image is being activated for the first time. This class should acquire
+        //! an image from the pool, configured for the provided descriptor. This may involve aliasing
+        //! from a heap, or simple object pooling. Images are held in an internal cache.
+        virtual MultiDeviceImage* ActivateImage(const TransientImageDescriptor& descriptor);
+
+        //! Called when an buffer is being activated for the first time. This class should acquire
+        //! an buffer from the pool, configured for the provided descriptor. This may involve aliasing
+        //! from a heap, or simple object pooling. Buffers are held in an internal cache.
+        virtual MultiDeviceBuffer* ActivateBuffer(const TransientBufferDescriptor& descriptor);
+
+        //! Called when a buffer is being de-allocated from the pool. Called during the last scope the attachment
+        //! is used, after all allocations for that scope have been processed. It will also be removed from the cache.
+        virtual void DeactivateBuffer(const AttachmentId& attachmentId);
+
+        //! Called when a image is being de-allocated from the pool. Called during the last scope the attachment
+        //! is used, after all allocations for that scope have been processed. It will also be removed from the cache.
+        virtual void DeactivateImage(const AttachmentId& attachmentId);
+
+        //! Called when all allocations for the current scope have completed.
+        void EndScope();
+
+        //! Called when the allocations / deallocations have completed for all scopes.
+        void End();
+
+        //! Get statistics for the pool (built during End).
+        AZStd::unordered_map<int, TransientAttachmentStatistics> GetStatistics() const;
+
+        //! Get pool descriptor
+        const TransientAttachmentPoolDescriptor& GetDescriptor() const;
+
+        //! Get the compile flags being used during the allocation of resources.
+        TransientAttachmentPoolCompileFlags GetCompileFlags() const;
+
+    private:
+        bool ValidateInitParameters(const TransientAttachmentPoolDescriptor& descriptor) const;
+
+        //! Remove the entry related to the provided attachmentId from the cache as it is probably stale now
+        void RemoveFromCache(RHI::AttachmentId attachmentId);
+
+        Scope* m_currentScope = nullptr;
+        TransientAttachmentPoolDescriptor m_descriptor;
+        TransientAttachmentPoolCompileFlags m_compileFlags = TransientAttachmentPoolCompileFlags::None;
+
+        //! MultiDeviceImage/Buffers added as attachments to scopes are tracked in an internal cache
+        ObjectCache<MultiDeviceResource> m_cache;
+
+        //! This map is used to reverse look up resource hash so we can clear them out of m_cache
+        //! once they have been replaced with a new resource at a different place in the heap.
+        AZStd::unordered_map<AttachmentId, HashValue64> m_reverseLookupHash;
```

Comment:
It's not clear why MD version of transient attachment pool needs m_cache/m_reverseLookupHash. It seems like it was copied over from Aliased Heap but not fully thought through. Activate < Resource > /Deactivate < Resource > are called multiple times by the FrameGraphCompiler in order to figure out which aliased resources overlap and which dont. The ones that do not overlap are allowed to alias. The cache within AliasedHeap exists to ensure we are not re-creating aliased resources multiple times and removing ones that will be replaced.  I guess my question is do we even need to cache MultiDeviceResource here?

Reply:
If activate and deactivate get called multiple times we will be adding/removing from the cache multiple times in a frame. At the very least we should remove RemoveFromCache from Deactivate < Resource > calls

Reply:
As @akioCL said, due to the API returning raw pointers, we have to hold on to the memory of the `MultiDeviceBuffer` or `MultiDeviceImage`, otherwise the memory would be cleaned up upon exiting the function and return an invalid pointer. 
Your point with removing the calls in the `Deactivate*` calls makes sense, but would we still need the ability to remove attachments from the cache?

Reply:
Hmm. Dont have a good solution atm. Maybe just clear the cache at Shutdown for now although this pool will maybe stay alive during the app's lifetime. AliasedHeap suffers the same issue to some degree but the cache does get cleared as heap pages get de-allocated eventually with the exception of the first page. Maybe FrameGraphCompiler can call and api based on some event or criteria which is not clear at the time.

Reply:
Have changed it thus for now, we'll keep it in mind then for the integration later if there is something better that can be done.